### PR TITLE
feat: added package installation progress

### LINKED ISSUE #1049: feat: Install Progress
### Is your feature request related to a problem? Please describe.

When using the installer provided by the tool, there is no way to show what is is busy doing. This leads me to not understanding how long it still needs, how big it is or what it is busy doing.

### Describe the solution you'd like to see

Just a simple progress bar with the packages install progress or even just the output from the installer, pnpm, npm, yarn ect.

### Describe alternate solutions

Pass in a flag at the start that will display it at the bottom or open up a terminal with the command already typed in

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by PiotrekPKP**
Closes #1049

## ✅ Checklist

- [x] I have followed every step in the [contributing guide](https://github.com/t3-oss/create-t3-app/blob/main/CONTRIBUTING.md) (updated 2022-10-06).
- [x] The PR title follows the convention we established [conventional-commit](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] I performed a functional test on my final commit

---

## Changelog

- added package installation progress to the cli working for all 3 supported package managers
- `npm` - output is inherited from `stderr`
- `pnpm` and `yarn` - output is redirected from `stdout`

---

### `npm`
image

### `pnpm`
image

### `yarn`
image

💯

### CODE REVIEW

**Comment by juliusmarminge** [comment]
Can we make the pnpm output prettier and not pushed so far to the right?

CleanShot 2023-01-28 at 11 52 39

File: cli/src/helpers/installDependencies.ts [hunk 1]
```diff
@@ -1,27 +1,76 @@
...
+    case "pnpm":
+      const pnpmSpinner = ora("Running pnpm install...").start();
+      const pnpmSubprocess = execa(pkgManager, ["install"], {
+        cwd: projectDir,
+        stdout: "pipe",
+      });
+
+      await new Promise<void>((res, rej) => {
+        pnpmSubprocess.stdout?.on("data", (data: Buffer) => {
+          pnpmSpinner.text = data.toString();
```
Comment (juliusmarminge) [inline cli/src/helpers/installDependencies.ts:38]: Something like that but maybe bit smarter matching idk?

```suggestion
pnpmSpinner.text = "Progress" + data.toString().split("Progress")[1];
```

**Comment by PiotrekPKP** [comment]
> Can we make the pnpm output prettier and not pushed so far to the right?
> 
> CleanShot 2023-01-28 at 11 52 39

`pnpm` uses this kind of output by default:
image

However, I think we can try to either:
- parse the progress bar manually and show it somewhat differently
- cut the string to the `|` sign and only show the text with progress info

What do you think is the better option? Of course the second one is easier haha

File: cli/src/helpers/installDependencies.ts (hunk 1 above)
Comment (PiotrekPKP) [reply cli/src/helpers/installDependencies.ts:38]: Yeah, that's what I've just commented haha

Comment (juliusmarminge) [reply cli/src/helpers/installDependencies.ts:38]: Astro's was clean and only reused a single line too, how does that work here?

Comment (PiotrekPKP) [reply cli/src/helpers/installDependencies.ts:38]: Tbf I'm not sure how their code works, I did the same thing they did. In Astro's codebase they also just use the data from the stdout :thinking:

Comment (juliusmarminge) [reply cli/src/helpers/installDependencies.ts:38]: > Tbf I'm not sure how their code works, I did the same thing they did. In Astro's codebase they also just use the data from the stdout :thinking: 

Maybe split by `\n`, take the last snd then spöit by the `|`?

**Comment by PiotrekPKP** [comment]
It looks like that now:
image

File: cli/src/helpers/installDependencies.ts (hunk 1 above)
Comment (PiotrekPKP) [reply cli/src/helpers/installDependencies.ts]: I used the split by `|` for now, looks as it should I think

File: cli/src/helpers/installDependencies.ts [hunk 2]
```diff
@@ -1,27 +1,78 @@
...
+    case "pnpm":
+      const pnpmSpinner = ora("Running pnpm install...").start();
+      const pnpmSubprocess = execa(pkgManager, ["install"], {
+        cwd: projectDir,
+        stdout: "pipe",
+      });
+
+      await new Promise<void>((res, rej) => {
+        pnpmSubprocess.stdout?.on("data", (data: Buffer) => {
+          if (data.toString().includes("Progress")) {
```
Comment (juliusmarminge) [inline cli/src/helpers/installDependencies.ts:38]:
```suggestion
          const text = data.toString();
          if (text.includes("Progress") && text.includes("|")) {
```

Seems like not every pnpm version uses the `|` so we should check for that too:

CleanShot 2023-01-28 at 12 19 32

Comment (juliusmarminge) [reply cli/src/helpers/installDependencies.ts:38]: https://user-images.githubusercontent.com/51714798/215263845-fd219e4a-1988-44ab-8298-80fba602d0ab.mp4

I for example doesn't get any of the progress output here

Comment (PiotrekPKP) [reply cli/src/helpers/installDependencies.ts]: I'm still checking for the progress text for older versions, no reason to show lines like `virtual store is at:` imo. Thanks for the info tho, I've updated the pr :+1:

Comment (juliusmarminge) [reply cli/src/helpers/installDependencies.ts]: CleanShot 2023-01-28 at 12 26 10

nice

**Review by juliusmarminge — APPROVED** [review]
Awesome work!

**Pull request merged**